llama.cpp

```diff
@@ -3779,3 +3794,43 @@ const char * llama_print_system_info(void) {
 const std::vector<std::pair<std::string, struct ggml_tensor *>>& llama_internal_get_tensor_map(struct llama_context * ctx) {
     return ctx->model.tensors_by_name;
 }
+
+
+void llama_log_set(llama_log_callback log_callback, void * user_data) {
+    g_state.log_callback = log_callback ? log_callback : llama_log_callback_default;
+    g_state.log_callback_user_data = user_data;
+}
+
+#if defined(_MSC_VER) && !defined(vsnprintf)
+#define vsnprintf _vsnprintf
+#endif
+
+static void llama_log_internal_v(llama_log_level level, const char * format, va_list args) {
+    va_list args_copy;
+    va_copy(args_copy, args);
+    char buffer[128];
+    int len = vsnprintf(buffer, 128, format, args);
+    if (len < 128) {
+        g_state.log_callback(level, buffer, g_state.log_callback_user_data);
+    } else {
+        char* buffer2 = new char[len+1];
+        vsnprintf(buffer2, len+1, format, args_copy);
+        buffer2[len] = 0;
+        g_state.log_callback(level, buffer2, g_state.log_callback_user_data);
+        delete[] buffer2;
+    }
+    va_end(args_copy);
+}
+
+static void llama_log_internal(llama_log_level level, const char * format, ...) {
+    va_list args;
+    va_start(args, format);
+    llama_log_internal_v(level, format, args);
+    va_end(args);
+}
+
+static void llama_log_callback_default(llama_log_level level, const char * text, void * user_data) {
+    (void) level;
+    (void) user_data;
+    fprintf(stderr, "%s\n", text);
```

Comment:
No default new-lines - this is bad practice

```suggestion
    fprintf(stderr, "%s", text);
```

Add `fflush(stderr);` after the `fprintf` to make sure things are displayed when a new-line is missing

Reply:
This will also solve the dilemma with the progress report

Reply:
I am not sure that it is necessarily a bad practice, it is pretty standard in logging libraries that the log granularity is at the line level.

Reply:
How about this: only print the progress updates if the user has not changed the logging function.

Reply:
We can do that, but the default new-line must be removed anyway. I am too used to writing it explicitly

Reply:
I was looking at what other logging libraries do, and for [go](https://pkg.go.dev/log) it says this:
>Every log message is output on a separate line: if the message being printed does not end in a newline, the logger will add one.

I think that could be a good compromise. I am concerned that if we allow printing partial lines with the logging function, it will make it hard to integrate it with virtually every other logging library.

Reply:
> We can do that, but the default new-line must be removed anyway. I am too used to writing it explicitly

This can make it difficult to log to some mediums, but it probably doesnt matter too much. I can do that.

Reply:
Alright I just changed the default logger callback not to output the new line at the end and flush the stream. The progress thing now also uses the logger. I also added a comment in llama.h explaining this.